ballerina/variable_validator.bal

```diff
@@ -0,0 +1,149 @@
+            if argument.isVariableDefinition() {
+                self.visitArgument(argument, data);
+            }
+        }
+        if fieldNode.getSelections().length() > 0 {
+            parser:Selection[] selections = fieldNode.getSelections();
+            foreach parser:Selection subSelection in selections {
+                self.visitSelection(subSelection, data);
+            }
+        } 
+    }
+
+    public isolated function visitFragment(parser:FragmentNode fragmentNode, anydata data = ()) {
+        foreach parser:Selection selection in fragmentNode.getSelections() {
+            self.visitSelection(selection, data);
+        }
+    }
+
+    public isolated function visitArgument(parser:ArgumentNode argumentNode, anydata data = ()) {
+        map<parser:VariableDefinition> variableDefinitions = <map<parser:VariableDefinition>> data;
+        string variableName = argumentNode.getVariableName();
+        parser:Location location = argumentNode.getValue().location;
+        if variableDefinitions.hasKey(variableName) {
+            parser:VariableDefinition variableDefinition = variableDefinitions.get(variableName);
+            self.visitedVariableDefinitions.push(variableName);
+            if self.variables.hasKey(variableName) {
+                argumentNode.setKind(getArgumentTypeKind(variableDefinition.kind));
+                Scalar|error value= <Scalar> self.variables.get(variableName);
+                if value is error {
+                    string message = string`Variable "$${variableName}" got invalid value ${self.variables.get(variableName).toString()};`;
+                    self.errors.push(getErrorDetailRecord(message, location));
+                } else {
+                    self.setArgumentValue(value, argumentNode, location);
+                }
+            } else if variableDefinition?.defaultValue != () {
+                argumentNode.setKind(getArgumentTypeKind(variableDefinition.kind));
+                Scalar defaultValue = <Scalar> (<parser:ArgumentValue> variableDefinitions.get(variableName)?.defaultValue).value;
+                self.setArgumentValue(defaultValue, argumentNode, location);
+            } else {
+                string message = string`Variable "$${variableName}" of required type ${variableDefinition.kind} was not provided.`;
+                self.errors.push(getErrorDetailRecord(message, location)); 
+            }
+        } else {
+            string message = string`Variable "$${variableName}" is not defined.`;
+            self.errors.push(getErrorDetailRecord(message, location)); 
+        }
+    }
+
+    public isolated function setArgumentValue(Scalar value, parser:ArgumentNode argument, parser:Location location) {
+        if argument.getKind() == parser:T_IDENTIFIER {
+            argument.setValue(value);
+        } else if (typeof value).toString().equalsIgnoreCaseAscii("typedesc " + getTypeName(argument)) {
```

Comment:
When comparing type names, let's not use `equalsIgnoreCase` function, since these names are case sensitive.

Shall we move retrieving the type name from the value to a separate function?

@MaryamZi aren't there any API to get the type name from a typedesc? If that's the case, shouldn't we have one to get type name from a typedesc value?